File: backend/app/services/serp_quota.py
"""Live SerpAPI / Serper search quota for the Individual Leads search UI.

SerpAPI: GET https://serpapi.com/account.json
Docs fields used: plan_searches_left, searches_per_month, this_month_usage,
total_searches_left, plan_name.

Never returns the API key. Short in-memory cache to avoid hammering the
account endpoint when the UI polls.
"""

from __future__ import annotations

import os
import time
from typing import Any

import httpx

SERPAPI_ACCOUNT_URL = "https://serpapi.com/account.json"
CACHE_TTL_S = 30.0
TIMEOUT_S = 12.0

_cache: dict[str, Any] = {"at": 0.0, "payload": None}


def _serp_config() -> dict[str, str] | None:
    serpapi = (os.getenv("SERPAPI_KEY") or os.getenv("SERP_API") or "").strip()
    if serpapi:
        return {"provider": "serpapi", "key": serpapi}
    serper = (os.getenv("SERPER_API_KEY") or "").strip()
    if serper:
        return {"provider": "serper", "key": serper}
    return None


def serp_provider_available() -> bool:
    return _serp_config() is not None


def _int_or_none(value: Any) -> int | None:
    if value is None or value == "":
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def _from_serpapi_account(data: dict[str, Any]) -> dict[str, Any]:
    """Normalize SerpAPI account.json into a UI-safe quota payload."""
    searches_per_month = _int_or_none(data.get("searches_per_month"))
    plan_left = _int_or_none(data.get("plan_searches_left"))
    total_left = _int_or_none(data.get("total_searches_left"))
    used = _int_or_none(data.get("this_month_usage"))
    extra = _int_or_none(data.get("extra_credits")) or 0
    renewal = _clean_date(data.get("plan_renewal_date"))

    # Prefer plan remaining; fall back to total (plan + extras).
    left = plan_left if plan_left is not None else total_left
    limit = searches_per_month
    if limit is None and left is not None and used is not None:
        limit = left + used
    if limit is None and left is not None:
        limit = left

    return {
        "available": True,
        "provider": "serpapi",
        "plan_name": (data.get("plan_name") or data.get("plan_id") or "").strip() or None,
        "searches_left": left,
        "searches_limit": limit,
        "searches_used": used,
        "extra_credits": extra if extra else None,
        "total_searches_left": total_left,
        "this_hour_searches": _int_or_none(data.get("this_hour_searches")),
        "plan_renewal_date": renewal,
        "label": _format_label(left, limit),
        "exhausted_message": _format_exhausted_message(renewal),
    }


def _clean_date(value: Any) -> str | None:
    raw = str(value or "").strip()
    if not raw:
        return None
    # SerpAPI returns YYYY-MM-DD; keep only the date part if a timestamp sneaks in.
    return raw[:10] if len(raw) >= 10 and raw[4] == "-" and raw[7] == "-" else raw


def _format_exhausted_message(renewal: str | None) -> str:
    if renewal:
        return f"No searches left for this month. Plan resets on {_friendly_date(renewal)}."
    return "No searches left for this month. Plan resets next month."


def _friendly_date(iso_date: str) -> str:
    """Format YYYY-MM-DD as 'Aug 16, 2026' without depending on locale."""
    months = (
        "Jan", "Feb", "Mar", "Apr", "May", "Jun",
        "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
    )
    try:
        year_s, month_s, day_s = iso_date.split("-", 2)
        month = int(month_s)
        day = int(day_s)
        if 1 <= month <= 12 and 1 <= day <= 31:
            return f"{months[month - 1]} {day}, {int(year_s)}"
    except (TypeError, ValueError):
        pass
    return iso_date


def _format_label(left: int | None, limit: int | None) -> str | None:
    if left is None:
        return None
    if limit is not None and limit >= 0:
        return f"{left}/{limit} searches left"
    return f"{left} searches left"


def _from_serper() -> dict[str, Any]:
    # Serper does not expose a public account/usage JSON comparable to SerpAPI.
    return {
        "available": True,
        "provider": "serper",
        "plan_name": None,
        "searches_left": None,
        "searches_limit": None,
        "searches_used": None,
        "extra_credits": None,
        "total_searches_left": None,
        "this_hour_searches": None,
        "plan_renewal_date": None,
        "label": "Serper search active",
        "exhausted_message": None,
        "note": "Live remaining counts require SerpAPI (serpapi.com).",
    }


def fetch_serp_quota(*, force: bool = False) -> dict[str, Any]:
    now = time.monotonic()
    if (
        not force
        and _cache["payload"] is not None
        and (now - float(_cache["at"])) < CACHE_TTL_S
    ):
        return dict(_cache["payload"])

    config = _serp_config()
    if not config:
        payload = {
            "available": False,
            "provider": None,
            "plan_name": None,
            "searches_left": None,
            "searches_limit": None,
            "searches_used": None,
            "extra_credits": None,
            "total_searches_left": None,
            "this_hour_searches": None,
            "plan_renewal_date": None,
            "label": None,
            "exhausted_message": None,
            "error": "SERPAPI_KEY (or SERPER_API_KEY) is not configured",
        }
        _cache["at"] = now
        _cache["payload"] = payload
        return dict(payload)

    if config["provider"] == "serper":
        payload = _from_serper()
        _cache["at"] = now
        _cache["payload"] = payload
        return dict(payload)

    try:
        with httpx.Client(timeout=TIMEOUT_S) as client:
            res = client.get(SERPAPI_ACCOUNT_URL, params={"api_key": config["key"]})
        if res.status_code in (401, 403):
            payload = {
                "available": False,
                "provider": "serpapi",
                "label": None,
                "error": "SerpAPI rejected the API key",
                "searches_left": None,
                "searches_limit": None,
                "searches_used": None,
            }
        elif res.status_code >= 400:
            payload = {
                "available": False,
                "provider": "serpapi",
                "label": None,
                "error": f"SerpAPI account error {res.status_code}",
                "searches_left": None,
                "searches_limit": None,
                "searches_used": None,
            }
        else:
            data = res.json() if res.content else {}
            if not isinstance(data, dict):
                raise ValueError("Unexpected SerpAPI account payload")
            if data.get("error"):
                payload = {
                    "available": False,
                    "provider": "serpapi",
                    "label": None,
                    "error": str(data.get("error")),
                    "searches_left": None,
                    "searches_limit": None,
                    "searches_used": None,
                }
            else:
                payload = _from_serpapi_account(data)
    except Exception as exc:  # noqa: BLE001 — surface soft failure to UI
        payload = {
            "available": False,
            "provider": "serpapi",
            "label": None,
            "error": str(exc)[:180],
            "searches_left": None,
            "searches_limit": None,
            "searches_used": None,
        }

    _cache["at"] = now
    _cache["payload"] = payload
    return dict(payload)


def invalidate_serp_quota_cache() -> None:
    _cache["at"] = 0.0
    _cache["payload"] = None

File: backend/src/serp-quota.js
const SERPAPI_ACCOUNT_URL = 'https://serpapi.com/account.json';
const CACHE_TTL_MS = 30_000;

let cache = { at: 0, payload: null };

function serpConfig() {
  const serpapi = (process.env.SERPAPI_KEY || process.env.SERP_API || '').trim();
  if (serpapi) return { provider: 'serpapi', key: serpapi };
  const serper = (process.env.SERPER_API_KEY || '').trim();
  if (serper) return { provider: 'serper', key: serper };
  return null;
}

function intOrNull(value) {
  if (value === null || value === undefined || value === '') return null;
  const n = Number(value);
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

function cleanDate(value) {
  const raw = String(value || '').trim();
  if (!raw) return null;
  return raw.length >= 10 && raw[4] === '-' && raw[7] === '-' ? raw.slice(0, 10) : raw;
}

function friendlyDate(isoDate) {
  const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(isoDate);
  if (!match) return isoDate;
  const month = Number(match[2]);
  const day = Number(match[3]);
  if (month < 1 || month > 12 || day < 1 || day > 31) return isoDate;
  return `${months[month - 1]} ${day}, ${Number(match[1])}`;
}

function formatExhaustedMessage(renewal) {
  if (renewal) return `No searches left for this month. Plan resets on ${friendlyDate(renewal)}.`;
  return 'No searches left for this month. Plan resets next month.';
}

function formatLabel(left, limit) {
  if (left === null) return null;
  if (limit !== null && limit >= 0) return `${left}/${limit} searches left`;
  return `${left} searches left`;
}

function fromSerpapiAccount(data) {
  const searchesPerMonth = intOrNull(data.searches_per_month);
  const planLeft = intOrNull(data.plan_searches_left);
  const totalLeft = intOrNull(data.total_searches_left);
  const used = intOrNull(data.this_month_usage);
  const extra = intOrNull(data.extra_credits) || 0;
  const renewal = cleanDate(data.plan_renewal_date);
  const left = planLeft !== null ? planLeft : totalLeft;
  let limit = searchesPerMonth;
  if (limit === null && left !== null && used !== null) limit = left + used;
  if (limit === null && left !== null) limit = left;

  return {
    available: true,
    provider: 'serpapi',
    plan_name: String(data.plan_name || data.plan_id || '').trim() || null,
    searches_left: left,
    searches_limit: limit,
    searches_used: used,
    extra_credits: extra || null,
    total_searches_left: totalLeft,
    this_hour_searches: intOrNull(data.this_hour_searches),
    plan_renewal_date: renewal,
    label: formatLabel(left, limit),
    exhausted_message: formatExhaustedMessage(renewal),
  };
}

export function invalidateSerpQuotaCache() {
  cache = { at: 0, payload: null };
}

export async function fetchSerpQuota({ force = false } = {}) {
  const now = Date.now();
  if (!force && cache.payload && now - cache.at < CACHE_TTL_MS) {
    return { ...cache.payload };
  }

  const config = serpConfig();
  if (!config) {
    const payload = {
      available: false,
      provider: null,
      plan_name: null,
      searches_left: null,
      searches_limit: null,
      searches_used: null,
      extra_credits: null,
      total_searches_left: null,
      this_hour_searches: null,
      plan_renewal_date: null,
      label: null,
      exhausted_message: null,
      error: 'SERPAPI_KEY (or SERPER_API_KEY) is not configured',
    };
    cache = { at: now, payload };
    return { ...payload };
  }

  if (config.provider === 'serper') {
    const payload = {
      available: true,
      provider: 'serper',
      plan_name: null,
      searches_left: null,
      searches_limit: null,
      searches_used: null,
      extra_credits: null,
      total_searches_left: null,
      this_hour_searches: null,
      plan_renewal_date: null,
      label: 'Serper search active',
      exhausted_message: null,
      note: 'Live remaining counts require SerpAPI (serpapi.com).',
    };
    cache = { at: now, payload };
    return { ...payload };
  }

  try {
    const url = new URL(SERPAPI_ACCOUNT_URL);
    url.searchParams.set('api_key', config.key);
    const res = await fetch(url, { signal: AbortSignal.timeout(12_000) });
    if (res.status === 401 || res.status === 403) {
      const payload = {
        available: false,
        provider: 'serpapi',
        label: null,
        error: 'SerpAPI rejected the API key',
        searches_left: null,
        searches_limit: null,
        searches_used: null,
      };
      cache = { at: now, payload };
      return { ...payload };
    }
    if (!res.ok) {
      const payload = {
        available: false,
        provider: 'serpapi',
        label: null,
        error: `SerpAPI account error ${res.status}`,
        searches_left: null,
        searches_limit: null,
        searches_used: null,
      };
      cache = { at: now, payload };
      return { ...payload };
    }
    const data = await res.json();
    if (data?.error) {
      const payload = {
        available: false,
        provider: 'serpapi',
        label: null,
        error: String(data.error),
        searches_left: null,
        searches_limit: null,
        searches_used: null,
      };
      cache = { at: now, payload };
      return { ...payload };
    }
    const payload = fromSerpapiAccount(data);
    cache = { at: now, payload };
    return { ...payload };
  } catch (error) {
    const payload = {
      available: false,
      provider: 'serpapi',
      label: null,
      error: String(error?.message || error).slice(0, 180),
      searches_left: null,
      searches_limit: null,
      searches_used: null,
    };
    cache = { at: now, payload };
    return { ...payload };
  }
}
